source/ada/lsp-ada_documents.adb

```diff
@@ -677,15 +708,141 @@ package body LSP.Ada_Documents is
    end Get_Decl_Kind;
 
+   -----------------
+   -- Get_Profile --
+   -----------------
+
+   function Get_Profile
+     (Node : Libadalang.Analysis.Basic_Decl) return LSP.Types.LSP_String
+   is
+      use Ada.Strings.Wide_Wide_Unbounded;
+      use Libadalang.Analysis;
+      use Libadalang.Common;
+
+      function To_Text (Node : Ada_Node'Class) return Wide_Wide_String;
+      --  Retrieve the node text and format it
+
+      function To_Profile
+        (Node : Libadalang.Analysis.Subp_Spec'Class)
+         return LSP.Types.LSP_String;
+
+      -------------
+      -- To_Text --
+      -------------
+
+      function To_Text (Node : Ada_Node'Class) return Wide_Wide_String
+      is
+         Node_Text : constant String :=
+           Langkit_Support.Text.To_UTF8 (Node.Text);
+         Result    : String  := Node_Text;
+         Was_Space : Boolean := False;
+         Cur       : Integer := Node_Text'First;
+      begin
+         for I in Node_Text'Range loop
+            if Node_Text (I) = ' ' then
+               if not Was_Space then
+                  Result (Cur) := Node_Text (I);
+                  Cur := Cur + 1;
+               end if;
+               Was_Space := True;
+            elsif Node_Text (I) /= ASCII.LF then
+               Was_Space := False;
+               Result (Cur) := Node_Text (I);
+               Cur := Cur + 1;
+            end if;
+         end loop;
+         return Ada.Characters.Conversions.To_Wide_Wide_String
+           (Result (Result'First .. Cur - 1));
+      end To_Text;
+
+      ----------------
+      -- To_Profile --
+      ----------------
+
+      function To_Profile
+        (Node : Libadalang.Analysis.Subp_Spec'Class)
+         return LSP.Types.LSP_String
+      is
+         Result  : Unbounded_Wide_Wide_String;
+         Params  : constant Param_Spec_Array := Node.P_Params;
+         Returns : constant Type_Expr := Node.F_Subp_Returns;
+      begin
+         if Params'Length > 0 then
+            Append (Result, "(");
+         end if;
+
+         for Param of Params loop
+            declare
+               Names : constant Defining_Name_List := Param.F_Ids;
+               Init  : constant Expr := Param.F_Default_Expr;
+               Item  : Unbounded_Wide_Wide_String;
+            begin
+               Append (Item, " :");
+
+               case Param.F_Mode is
+                  when Ada_Mode_Default | Ada_Mode_In =>
+                     Append (Item, " in ");
+                  when Ada_Mode_In_Out =>
+                     Append (Item, " in out ");
+                  when Ada_Mode_Out =>
+                     Append (Item, " out ");
+               end case;
+
+               Append (Item, To_Text (Param.F_Type_Expr));
+
+               if not Init.Is_Null then
+                  Append (Item, " := ");
+                  Append (Item, To_Text (Init));
+               end if;
+
+               for J in Names.First_Child_Index .. Names.Last_Child_Index loop
```

Comment:
Is it really needed? Why can't we just display the subprogram declaration like for tooltips?

Reply:
The tooltips displaid a multi-lines profile while we want a pretty-printed single line profile.
Let's take the following declaration:

procedure Foo
   (I     : in out Integer;
    Bar : String);

The tooltip will show the declaration itself, while the pretty-printed single line profile is:

Foo (I : in out Integer; Bar : in String)

Reply:
Ok, fine by me then